### PR TITLE
Add RUBY_BUILD_TARBALL_OVERRIDE to override the ruby tarball URL

### PR DESCRIPTION
**Pull request by eregon**
* And add RUBY_BUILD_IGNORE_CHECKSUM when it is expected the alternative URL has different contents.

This is useful for instance when testing a truffleruby release which is not published yet.
Then I need to use a custom URL, like:
```
RUBY_BUILD_MIRROR_PACKAGE_URL=https://github.com/my/url.tar.gz ruby-build truffleruby-23.1.0 ~/.rubies/truffleruby-test
```
I also want this to try dev builds at a different URL, hence this change.
(and those might have a different checksum too, see below)

And finally I would also like a way to skip the checksum check, for instance when trying a variant of a release which therefore has a different URL and contents but yet I want to reuse the existing definition.
For example this can be useful if I want to try the CE build instead of GFTC build of truffleruby 23.1.0, like this:
```
RUBY_BUILD_IGNORE_CHECKSUM=1 RUBY_BUILD_MIRROR_PACKAGE_URL=https://github.com/oracle/truffleruby/releases/download/graal-23.1.0/truffleruby-community-23.1.0-linux-amd64.tar.gz ruby-build truffleruby-23.1.0 ~/.rubies/truffleruby-test
```

### CODE REVIEW

File: bin/ruby-build [hunk 1]
```diff
@@ -1194,7 +1198,7 @@ isolated_gem_install() {
 apply_ruby_patch() {
   local patchfile
   case "$1" in
-  ruby-* | jruby-* | rubinius-* | truffleruby-* )
+  ruby-* | jruby-* | mruby-* | picoruby-* | truffleruby-* | truffleruby+graalvm-* )
```
Comment (mislav) [inline bin/ruby-build:1201]: This feels like it belongs in a separate PR

**Review by mislav — COMMENTED** [review]
I'm skeptical of this feature because I already wasn't a fan of the original RUBY_BUILD_MIRROR_PACKAGE_URL feature, and this adds even more features around it https://github.com/rbenv/ruby-build/pull/1793

> And finally I would also like a way to skip the checksum check, for instance when trying a variant of a release which therefore has a different URL and contents but yet I want to reuse the existing definition.

From everything you described so far, it sounds like what you're doing isn't at all mirroring existing releases, but dynamically modifying the target that a build definition downloads and installs. If you need to modify a truffleruby definition slightly, would it be an option for you to create a custom definition and provide `ruby-build` with a path to your custom definition?

File: bin/ruby-build (hunk 1 above)
Comment (eregon) [reply bin/ruby-build:1201]: It's just a cleanup I noticed while working on this and already its own commit, but sure I can split it up.

**Comment by eregon** [comment]
Right, this is indeed not mirroring.
@mislav How would you feel about adding another environment variable then?
I was thinking something like `RUBY_BUILD_OVERRIDE_RUBY_URL` or so originally (and this would only override the URL if it's "a ruby" package, which then would use the same check as [here](https://github.com/rbenv/ruby-build/pull/2258#discussion_r1334230471)).

(then I saw RUBY_BUILD_MIRROR_PACKAGE_URL already existed and it seemed nicer to extend it)

**Comment by mislav** [comment]
I take this as custom build definitions not being a suitable workaround for you?

**Comment by eregon** [comment]
Yeah they would be too annoying to be of any real use.
Also for the first 2 use cases, those are actually quite related to mirroring, and custom definitions there would not be testing what we want.

**Comment by eregon** [comment]
@mislav What do you think of the above? Which way would be OK to merge?

File: bin/ruby-build (hunk 1 above)
Comment (mislav) [reply bin/ruby-build:1201]: The problem with this change being bundled here, apart from being wholly unrelated, is that it won't get its own changelog entry automatically upon cutting a release with `script/release` (nor when creating a release using the web interface and choosing "Generate release notes"). I believe that all feature changes deserve a changelog entry.

Comment (mislav) [inline bin/ruby-build:1201]:
```suggestion
  ruby-* | jruby-* | mruby-* | picoruby-* | truffleruby[+-]* )
```

File: bin/ruby-build [hunk 2]
```diff
@@ -365,14 +365,18 @@ fetch_tarball() {
...
     package_url="${package_url%%#*}"
+  fi
 
-    if [ -n "$RUBY_BUILD_MIRROR_URL" ]; then
-      if [[ -z "$RUBY_BUILD_DEFAULT_MIRROR" || $package_url != */cache.ruby-lang.org/* ]]; then
-        mirror_url="${RUBY_BUILD_MIRROR_URL}/$checksum"
-      fi
-    elif [ -n "$RUBY_BUILD_MIRROR_PACKAGE_URL" ]; then
-      mirror_url="$RUBY_BUILD_MIRROR_PACKAGE_URL"
+  if [ -n "$RUBY_BUILD_IGNORE_CHECKSUM" ]; then
```
Comment (mislav) [inline bin/ruby-build:370]: I do not believe we should expose this additional option, since we would not want anyone to actually use it.

Instead, in order to support the use-case that you are describing (testing Ruby releases that do not yet have their own definition), I suggest that we introduce a new environment variable completely orthogonal to the mirror feature (since this is technically not related to mirroring, either). The new environment variable could be called something like `RUBY_BUILD_TARBALL_OVERRIDE`, and the value read from it would replace both the URL and checksum values in `fetch_tarball`. If the environment variable only supplies a URL, then checksum checks would be off anyway, but a checksum could also be optionally supplied too. How does that sound?

**Review by mislav — CHANGES_REQUESTED** [review]

File: bin/ruby-build (hunk 2 above)
Comment (eregon) [reply bin/ruby-build:370]: Yes that sounds great, I'll update the PR with it.

**Comment by eregon** [comment]
@mislav Updated, could you review and merge if it's OK?

File: bin/ruby-build [hunk 3]
```diff
@@ -371,6 +371,10 @@ fetch_tarball() {
   local checksum
   local extracted_dir
 
+  if is_ruby_package "$1" && [ -n "$RUBY_BUILD_TARBALL_OVERRIDE" ]; then
```
Comment (eregon) [inline bin/ruby-build:374]: The reason we need the is_ruby_package is because fetch_tarball is also used for non-ruby packages like openssl and overriding both URLs with the same URL would make no sense.

File: bin/ruby-build (hunk 1 above)
Comment (eregon) [reply bin/ruby-build]: In the updated PR it's fully related and necessary.

**Comment by eregon** [comment]
Rebased on top of master to include https://github.com/rbenv/ruby-build/pull/2268, and locally `shellcheck bin/ruby-build` is clean.

**Review by mislav — APPROVED** [review]
Love the brevity of the final solution! Thank you.

**Comment by eregon** [comment]
Indeed it's beautiful how small it is, I was positively surprised as well :)
